### PR TITLE
fix: pretty printer: a tuple argument becomes multiple arguments #1384

### LINKED ISSUE #1384: pretty printer: a tuple argument becomes multiple arguments
```
method Main() {
  var f : ((int, int)) -> int := (x : (int, int)) => 1;
}
```

is pretty printed to:

```
method Main()
{
  var f: (int, int) -> int := (x: (int, int)) => 1;
}
```

This drops parentheses around `(int, int)`, turning a one argument function into a two arguments function.

### PR DESCRIPTION
**Pull request by prvshah51**
fix #1384

### CODE REVIEW

File: Source/Dafny/AST/DafnyAst.cs [hunk 1]
```diff
@@ -2506,10 +2506,11 @@ public ArrowType(IToken tok, ArrowTypeDecl atd, List<Type> typeArgs, Type result
         // arrows are right associative, so we need parentheses around the domain type
         domainNeedsParens = true;
       } else {
-        // if the domain type consists of a single tuple type, then an extra set of parentheses is needed
```
Comment (keyboardDrummer) [inline Source/Dafny/AST/DafnyAst.cs:2509]: "needed" seems correct, no?

File: Source/Dafny/AST/DafnyAst.cs [hunk 2]
```diff
@@ -2506,10 +2506,11 @@ public ArrowType(IToken tok, ArrowTypeDecl atd, List<Type> typeArgs, Type result
...
+        // if the domain type consists of a single tuple type, then an extra set of parentheses is need
         // Note, we do NOT call .AsDatatype or .AsIndDatatype here, because those calls will do a NormalizeExpand().  Instead, we do the check manually.
-        var udt = typeArgs[0].Normalize() as UserDefinedType;  // note, we do Normalize(), not NormalizeExpand(), since the TypeName will use any synonym
-        if (udt != null && udt.ResolvedClass is TupleTypeDecl) {
+        var udt = typeArgs[0].Normalize() as UserDefinedType;  // note, we do Normalize(), not NormalizeExpand(), since the TypeName will use any synony
+        
+        if (udt != null && BuiltIns.IsTupleTypeName(udt.FullName))  {
```
Comment (keyboardDrummer) [inline Source/Dafny/AST/DafnyAst.cs:2513]: The line break here seems a bit extreme, or do you prefer it?

**Review by keyboardDrummer — APPROVED** [review]

File: Source/Dafny/AST/DafnyAst.cs (hunk 2 above)
Comment (prvshah51) [reply Source/Dafny/AST/DafnyAst.cs:2513]: No, I commented the original code and wrote change in a new line and removed that line leaving the break before PR. I can change it.

File: Source/Dafny/AST/DafnyAst.cs (hunk 1 above)
Comment (prvshah51) [reply Source/Dafny/AST/DafnyAst.cs:2509]: I think it was changed by mistake.

File: Test/git-issues/git-issue-1384.dfy [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// RUN: %dafny /compile:0 /env:0 /dprint:- "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+method Test()
+{
+    
+    var f : ((int,int)) -> int := ( x: (int,int)) => 1;
+
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-1384.dfy:8]: For styling, use 2 spaces for indents, no space between an identifier and the following `:`.
```suggestion    
    var f: ((int,int)) -> int := (x: (int,int)) => 1;
```
(Unless the example is intentionally different from the pretty-printed output.)

File: Source/Dafny/AST/DafnyAst.cs [hunk 4]
```diff
@@ -2718,7 +2718,7 @@ public class UserDefinedType : NonProxyType {
 
     public string FullName {
       get {
-        if (ResolvedClass != null && !ResolvedClass.EnclosingModuleDefinition.IsDefaultModule) {
+        if (ResolvedClass != null && ResolvedClass.EnclosingModuleDefinition != null && !ResolvedClass.EnclosingModuleDefinition.IsDefaultModule) {
```
Comment (keyboardDrummer) [inline Source/Dafny/AST/DafnyAst.cs:2721]: Consider writing 

`ResolvedClass != null && ResolvedClass.EnclosingModuleDefinition != null && !ResolvedClass.EnclosingModuleDefinition.IsDefaultModule` 

as

`ResolvedClass?.EnclosingModuleDefinition?.IsDefaultModule == false`

**Review by keyboardDrummer — APPROVED** [review]

**Pull request merged**